### PR TITLE
MAINT: Introduce various linting and misc fixes to `numpy.typing`

### PR DESCRIPTION
**Pull request by BvB93**
Follow up on https://github.com/numpy/numpy/pull/19784#issuecomment-909357832.

This PR fixes a introduces a number of linting-related fixes to the `numpy.typing` subpackage.

While as a general rule of thumb linting PRs are not encouraged (for various reasons),
`numpy.typing` is both small and young enough that this is not a big concern.

Also, it provides a nice opportunity for the introduction of some other small changes that were on the to-do list,
such as [PEP 585](https://www.python.org/dev/peps/pep-0585/)-based subscription of builtin types and switching over a number of _de facto_ type-check-only 
files from the `.py` to `.pyi` extension.

### CODE REVIEW

File: tools/linter.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 # computing the diff itself.
 EXCLUDE = (
     "numpy/typing/tests/data/",
+    "numpy/typing/_char_codes.py",
```
Comment (BvB93) [inline tools/linter.py:17]: In `_char_codes.py` I strongly prefer the longer line-length to keep things more ordered,
hence why it's been excluded from the linting tests. 

https://github.com/numpy/numpy/blob/1cadccc6012124e38e02a78b39bf268123d272c1/numpy/typing/_char_codes.py#L81-L95

**Review by BvB93 — COMMENTED** [review]
Below is the only line that could still trigger a warning (as it consists of 84 characters).
Going from a single- to multi-line function definition doesn't look nearly as good in the docs here, in my opinion. Hence why it has remained unchanged. 

https://github.com/numpy/numpy/blob/1cadccc6012124e38e02a78b39bf268123d272c1/numpy/typing/__init__.py#L202-L203

**Review by rgommers — APPROVED** [review]
This looks great, thanks @BvB93. It's also more than style changes only, the PEP 585 and `.py` -> `.pyi` changes are useful.

**Comment by BvB93** [comment]
> It's also more than style changes only, the PEP 585 and .py -> .pyi changes are useful.

Yeah, that's fair. I've updated the title a bit to reflect this.

**Pull request merged**